src/main/java/org/mskcc/cmo/messaging/impl/JSGatewayImpl.java

```diff
@@ -0,0 +1,256 @@
+        exec.shutdownNow();
+        shutdownInitiated = true;
+        publishingShutdownLatch.await();
+        natsConnection.close();
+    }
+
+    /**
+     * Configure basic message handler logic.
+     * @param msg
+     * @param messageClass
+     * @param messageConsumer
+     */
+    public void onMessage(Message msg, Class messageClass, MessageConsumer messageConsumer) {
+        String payload = new String(msg.getData(), StandardCharsets.UTF_8);
+        Object message = null;
+
+        try {
+            message = mapper.readValue(payload, messageClass);
+        } catch (JsonProcessingException ex) {
+            LOG.error("Error deserializing NATS message: " + payload, ex);
+        }
+        if (message != null) {
+            messageConsumer.onMessage(message);
+        }
+    }
+
+    /**
+     * Writes to publishing logger file.
+     * @param subject
+     * @param message
+     */
+    public void writeToPublishingLoggerFile(String subject, String message) {
+        String currentDate = LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE);
+        StringBuilder builder = new StringBuilder();
+        builder.append(currentDate)
+                .append("\t")
+                .append(subject)
+                .append("\t")
+                .append(message)
+                .append("\n");
+
+        try {
+            fileUtil.writeToFile(publishingLoggerFile,
+                    builder.toString());
+        } catch (IOException ex) {
+            LOG.error("Error during attempt to log publishing task to logger file", ex);
+        }
+    }
+
+    private class NATSPublisher implements Runnable {
+        Connection natsConn;
+        JetStream jsConn;
+        boolean interrupted;
+
+        /**
+         * NATSPublisher constructor.
+         * @param options
+         * @throws IOException
+         * @throws InterruptedException
+         */
+        public NATSPublisher(Options options) throws IOException, InterruptedException {
+            this.natsConn = Nats.connect(options);
+            this.jsConn = natsConn.jetStream(JetStreamOptions.DEFAULT_JS_OPTIONS);
+            this.interrupted = Boolean.FALSE;
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    PublishingQueueTask task = publishingQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (task != null) {
+                        String msg = mapper.writeValueAsString(task.payload);
+                        try {
+                            PublishAck ack = jsConn.publish(task.subject, msg.getBytes());
+                            if (ack.getError() != null) {
+                                writeToPublishingLoggerFile(task.subject, msg);
+                            }
+                        } catch (Exception e) {
+                            if (msg == null) {
+                                writeToPublishingLoggerFile(task.subject, "message is null");
```

Comment:
we can revert this change that checks if the message is null because if it is then we'll see it in the logger.